src/udev/udev-spawn.c

```diff
@@ -54,35 +56,54 @@ static int on_spawn_io(sd_event_source *s, int fd, uint32_t revents, void *userd
         assert(fd == spawn->fd_stdout || fd == spawn->fd_stderr);
         assert(!spawn->result || spawn->result_len < spawn->result_size);
 
-        if (fd == spawn->fd_stdout && spawn->result) {
+        if (fd == spawn->fd_stdout && spawn->result && !spawn->truncated) {
+                /* When reading to the result buffer, use the maximum available size, to detect truncation. */
+                read_to_result = true;
                 p = spawn->result + spawn->result_len;
                 size = spawn->result_size - spawn->result_len;
+                assert(size > 0);
         } else {
+                /* When reading to the local buffer, keep the space for the trailing NUL. */
+                read_to_result = false;
                 p = buf;
-                size = sizeof(buf);
+                size = sizeof(buf) - 1;
```

Comment:
Suggest to let both reserve 1 byte
`        if (fd == spawn->fd_stdout && spawn->result && !spawn->truncated) {
                p = spawn->result + spawn->result_len;
                size = spawn->result_size - spawn->result_len;
                assert(size > 1);
        } else {
                p = buf;
                size = sizeof(buf);
        }
        size -= 1; /* Reserve for trailing NUL. */
  `
  to simply code, so that `l--` on line 99 is needless any more

Reply:
That's intentional.

When reading to `spawn->result`, we need to detect truncation. When the read size equals to the buffer size, then the result is truncated, as we need to insert trailing NUL. Hence, we need to pass the size of the buffer. Such technique is often used at other places.

When we read just for logging or draining, it is not necessary to detect truncation. Hence, pass -1 size to keep the trailing NUL.